src/gluonts/itertools.py

```diff
@@ -289,6 +289,65 @@ def __iter__(self):
         return filter(self.fn, self.iterable)
 
 
+@dataclass
+class WeightedIterables:
+    """
+    Given a list of Iterables `iterables`, generate samples from them.
+
+    When `probabilities` is given, sample iteratbles according to it.
+    When `probabilities` is not given, sample iterables uniformly.
+
+    When one iterable exhausts, the sampling probabilities for it will be set to 0.
+
+        >>> from toolz import take
+        >>> from gluonts.itertools import Cyclic
+        >>> a = [1, 2, 3]
+        >>> b = [4, 5, 6]
+        >>> it = iter(WeightedIterables([a, b], [1, 0]))
+        >>> list(take(5, it))
+        [1, 2, 3]
+
+
+        >>> a = [1, 2, 3]
+        >>> b = [4, 5, 6]
+        >>> it = iter(WeightedIterables([Cyclic(a), b], [1, 0]))
+        >>> list(take(5, it))
+        [1, 2, 3, 1, 2]
+
+    """
+
+    iterables: List[Iterable]
+    probabilities: Optional[List[float]] = None
+    random_state: RandomState = field(default_factory=RandomState)
+
+    def __post_init__(self):
+        if self.probabilities:
+            # copy the probabilities as we may modify it
+            self.probabilities = list(self.probabilities)
+        else:
+            self.probabilities = [1.0 / len(self.iterables)] * len(
+                self.iterables
+            )
+
+    def __iter__(self):
+        iterators = [iter(it) for it in self.iterables]
+        while True:
+            idx = self.random_state.choice(
+                range(len(iterators)), p=self.probabilities
+            )
+
+            try:
+                yield next(iterators[idx])
+            except StopIteration:
+                self.probabilities[idx] = 0
```

Comment:
This makes the object stateful: if while iterating the first time some probabilities are set to zero, they will remain zero the second time you iterate. I don't think this is what we want?

I think in this iterable approach we should have a local copy of `self.probabilities` in the scope of `__iter__`

Reply:
You are absolutely right, changed.